### DISCUSSION TITLE
How to test qute checked templates?

### DISCUSSION THREAD

**Discussion by jonafll** [quarkusio/quarkus#27201] [Q&A] [answered]
Hi,

I want to validate the rendered xml of my checked templates without starting Quarkus. But when I run my tests, an UnsatisfiedLinkError exception is thrown. 

According to my [research](https://www.baeldung.com/java-native#2-testing-native-methods) I need to load a system library for native methods. Unfortunately, there are no examples of this type of testing in any of the Quarkus Qute Guides.

Here is my example code:

```java
package de.web.pdf.qute;

import io.quarkus.qute.CheckedTemplate;
import io.quarkus.qute.TemplateInstance;
import lombok.Data;

@CheckedTemplate
public class PersonTemplate {

    @Data
    public static class PersonData {
        String name;
        Integer age;
    }

    public static native TemplateInstance person(PersonTemplate.PersonData data);
}
```

```java
package de.web.pdf.qute;

import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import java.util.Optional;

@ApplicationScoped
public class PdfXmlRenderer {

    public String buildPerson(String name, Integer age) {
        PersonTemplate.PersonData data = new PersonTemplate.PersonData();
        data.name = name;
        data.age = age;

        return PersonTemplate.person(data).render();
    }
}
```

```java
package de.web.pdf.qute;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import javax.inject.Inject;

public class PdfXmlRendererTest {

    @Test
    public void test_person_xml() {

        String name = "John";
        Integer age = 24;

        PdfXmlRenderer pdfXmlRenderer = new PdfXmlRenderer();
        String personXml = pdfXmlRenderer.buildPerson(name, age); // This throws an UnsatisfiedLinkError

        // Assertions...
    }
}
```

[a comment by an automated account is not shown]

**Comment by mkouba** [comment, marked as answer]
You can't test qute checked templates in a unit test. Instead, you'll need a JUnit5 extension that starts the quarkus application, i.e. annotated the test class with the `@io.quarkus.test.junit.QuarkusTest` annotation. I'd recommend you to start with our [Testing Your Application](https://quarkus.io/guides/getting-started-testing) guide.

**Reply by rodcheater** [reply]
> You can't test qute checked templates in a unit test.

Shouldn’t this be addressed? Unit testing is pretty fundamental to modern software development. I don’t really want to be messing around with HtmlUnit and a full-blown instance of my application to be able to test a single resource class.

**Reply by mkouba** [reply]
> to be able to test a single resource class.

Well, if you don't start the full stack then you're basically testing some mocks. We don't have a way to mock type-safe templates ATM. I know that Panache has some [crazy mocking support](https://quarkus.io/guides/hibernate-orm-panache#mocking), we could add something similar, but I'm not sure it's worth the hassle. I mean, we would not be able to run the validation etc.

The situation could be better with [template records](https://quarkus.io/guides/qute-reference#template-records).

**Reply by rodcheater** [reply]
I rediscovered template records in the documentation just after I sent my message. I was really excited to try them out when I got to work this morning, only to find out that template records don’t support fragments. 🙄

**Reply by mkouba** [reply]
https://github.com/quarkusio/quarkus/pull/46985

**Reply by rodcheater** [reply]
I’ve also sent https://github.com/quarkusio/quarkus/pull/47001 to address another issue which is blocking me. 🙏
